hoomd/hpmc/external/user.py

```diff
@@ -0,0 +1,142 @@
+# License.
+
+import hoomd
+from hoomd import _compile
+from hoomd.hpmc import integrate
+from hoomd.hpmc import _jit
+from hoomd.operation import _HOOMDBaseObject
+from hoomd.logging import log
+
+
+class CPPExternalField(_HOOMDBaseObject):
+    r'''Define an external field imposed on all particles in the system.
+
+    Args:
+        code (str): C++ function body to compile.
+        clang_exec (str, optional): The clang executable to use, defaults to
+        ``'clang'``.
+
+    Potentials added using external.CPPExternalField are added to the total
+    energy calculation in :py:mod:`hpmc <hoomd.hpmc>` integrators. The
+    :py:class:`CPPExternalField` external field takes C++ code, JIT compiles it
+    at run time and executes the code natively in the MC loop with full
+    performance. It enables researchers to quickly and easily implement custom
+    energetic field intractions without the need to modify and recompile HOOMD.
+
+    .. rubric:: C++ code
+
+    Supply C++ code to the *code* argument and :py:class:`CPPExternalField` will
+    compile the code and call it to evaluate forces. Compilation assumes that a
+    recent ``clang`` installation is on your PATH.
+
+    The text provided in *code* is the body of a function with the following
+    signature:
+
+    .. code::
+
+        float eval(const BoxDim& box,
+        unsigned int type_i,
+        const vec3<Scalar>& r_i,
+        const quat<Scalar>& q_i
+        Scalar diameter,
+        Scalar charge
+        )
+
+    * ``vec3`` and ``quat`` are defined in HOOMDMath.h.
+    * *box* is the system box.
+    * *type_i* is the particle type.
+    * *r_i* is the particle position
+    * *q_i* the particle orientation.
+    * *diameter* the particle diameter.
+    * *charge* the particle charge.
+    * Your code *must* return a value.
+
+    Example:
+
+    .. code-block:: python
+
+        gravity = "return r_i.z + box.getL().z/2;"
+        external = hoomd.jit.external.CPPExternalField(code=gravity)
+
+    Note:
+        CPPExternalField does not support execution on GPUs.
+    '''
+    _integrator_pairs = {
+        integrate.Sphere: _jit.ExternalFieldJITSphere,
+        integrate.ConvexPolygon: _jit.ExternalFieldJITConvexPolygon,
+        integrate.SimplePolygon: _jit.ExternalFieldJITSimplePolygon,
+        integrate.ConvexPolyhedron: _jit.ExternalFieldJITConvexPolyhedron,
+        integrate.ConvexSpheropolyhedron: _jit.ExternalFieldJITSpheropolyhedron,
+        integrate.Ellipsoid: _jit.ExternalFieldJITEllipsoid,
+        integrate.ConvexSpheropolygon: _jit.ExternalFieldJITSpheropolygon,
+        integrate.FacetedEllipsoid: _jit.ExternalFieldJITFacetedEllipsoid,
+        integrate.Polyhedron: _jit.ExternalFieldJITPolyhedron,
+        integrate.Sphinx: _jit.ExternalFieldJITSphinx
+    }
+
+    def __init__(self, code, clang_exec='clang'):
+        self._llvm_ir = _compile.to_llvm_ir(self._wrap_cpu_(code))
```

Comment:
Actually, the situation here is worse than I thought. In MPI simulations, the `fork()` to run `clang` causes problems with MPI and/or CUDA. I haven't narrowed down exactly what conditions are needed for a minimal reproducer but the behavior we get is this with 2-rank GPU simulations:
- On `brett`: forking, even to just run `ls` corrupts the particle data memory leading to random errors in `MPI_ISend`, errors that particles are missing when gathering in `getSnapshot`, or `cudaErrorIllegalAddress` errors.
- On `summit`: the fork hangs on both ranks, even when running just `ls`.

We are currently working around this by pre-compiling and specifying the LLVM IR when running. I'm going to push a commit that enables this for now so we have something that works.

Longer term, we should implement a better solution. Options include:
1) Identify the root cause of the failures and work around them somehow to enable forking.
2) Use clang as a library to avoid forking altogether.
3) Accept the llvm IR specification as a solution going forward.
4) Any other ideas?

Reply:
Resolving now that Josh figured out how to use clang as a library.